pkg/ipam/node.go

```diff
@@ -794,11 +796,14 @@ func (n *Node) MaintainIPPool(ctx context.Context) error {
 
 // Update cilium node IPAM status with excess IP release data
 func (n *Node) PopulateIPReleaseStatus(node *v2.CiliumNode) {
+	// maintainIPPool() might not have run yet since the last update from agent.
+	// Attempt to remove any stale entries
+	n.removeStaleReleaseIPs()
```

Comment:
This seems to introduce a data race, (go test -race inside `pkg/ipam`):

````
WARNING: DATA RACE
Read at 0x00c00097e3c0 by goroutine 120:
  runtime.mapiterinit()
      /usr/lib/go/src/runtime/map.go:802 +0x0
  github.com/cilium/cilium/pkg/ipam.(*Node).removeStaleReleaseIPs()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/ipam/node.go:577 +0x8c
  github.com/cilium/cilium/pkg/ipam.(*Node).PopulateIPReleaseStatus()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/ipam/node.go:801 +0x52
  github.com/cilium/cilium/pkg/ipam.(*IPAMSuite).TestNodeManagerReleaseAddress.func5()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/ipam/node_manager_test.go:446 +0xaf

Previous write at 0x00c00097e3c0 by goroutine 60:
  runtime.mapassign_faststr()
      /usr/lib/go/src/runtime/map_faststr.go:202 +0x0
  github.com/cilium/cilium/pkg/ipam.(*Node).maintainIPPool()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/ipam/node.go:679 +0x2292
  github.com/cilium/cilium/pkg/ipam.(*Node).MaintainIPPool()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/ipam/node.go:784 +0x130
  github.com/cilium/cilium/pkg/ipam.(*NodeManager).Update.func2()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/ipam/node_manager.go:273 +0x7d
  github.com/cilium/cilium/pkg/trigger.(*Trigger).waiter()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:198 +0x35b
  github.com/cilium/cilium/pkg/trigger.NewTrigger·dwrap·1()
      /home/gandro/Isovalent/go/src/github.com/cilium/cilium/pkg/trigger/trigger.go:119 +0x39
```

I think we need to revisit the locking for the `ipReleaseStatus` map. We seem to be reading it without the mutex, but writing to it with the mutex held, which I don't think is safe.

Reply:
@gandro Thanks for catching it, pushed an update to `removeStaleReleaseIPs()` func.